src/more/more.rs

```diff
@@ -137,30 +179,74 @@ fn reset_term(term: &mut redox_termios::Termios) {
 fn more(matches: getopts::Matches) {
     let files = matches.free;
     let mut f = File::open(files.first().unwrap()).unwrap();
-    let mut buffer = [0; 1024];
 
+    let mut buffer = [0; 1024];
     let mut term = setup_term();
+    let (term_columns, term_lines) = match term_size::dimensions() {
+        Some((w, h)) => (w, h-1), // -1 to leave space for prompt
+        None => (80, 30)
+    };
+
+    let mut want_lines = term_lines; // start with a full page; count down
+    let mut columns = term_columns;   // for consistency, count down
 
-    let mut end = false;
-    while let Ok(sz) = f.read(&mut buffer) {
-        if sz == 0 {
+    'chunks: while let Ok(size) = f.read(&mut buffer) {
+        if size == 0 {
             break;
         }
-        stdout().write(&buffer[0..sz]).unwrap();
-        for byte in std::io::stdin().bytes() {
-            match byte.unwrap() {
-                b' ' => break,
-                b'q' | 27 => {
-                    end = true;
-                    break;
+        let mut write_start = 0;    // start of next write
+        let mut point = 0;          // next char when counting lines
+
+        loop {
+            // find a subrange with the right number of lines
+            while want_lines > 0 {
+                let c = buffer[point];
```

Comment:
This will bounds check, so if you can, it may be better to rewrite this loop using `slice::split` or something like that.